newIDE/app/src/UI/SortableVirtualizedItemList/index.js

```diff
@@ -100,73 +100,77 @@ export default class SortableVirtualizedItemList<Item> extends React.Component<
-                          const dropTarget = connectDropTarget(
-                            <div>
-                              {isOver && <DropIndicator canDrop={canDrop} />}
-                              <ItemRow
-                                item={item}
-                                itemName={itemName}
-                                isBold={isItemBold ? isItemBold(item) : false}
-                                onRename={newName =>
-                                  this.props.onRename(item, newName)
-                                }
-                                editingName={nameBeingEdited}
-                                getThumbnail={
-                                  getItemThumbnail
-                                    ? () => getItemThumbnail(item)
-                                    : undefined
-                                }
-                                selected={selectedItems.indexOf(item) !== -1}
-                                onItemSelected={this.props.onItemSelected}
-                                errorStatus={
-                                  erroredItems
-                                    ? erroredItems[itemName] || ''
-                                    : ''
-                                }
-                                buildMenuTemplate={() =>
-                                  this.props.buildMenuTemplate(item, index)
-                                }
-                                onEdit={onEditItem}
-                                hideMenuButton={windowWidth === 'small'}
-                                connectIconDragSource={
+                      {nameBeingEdited ? (
+                        itemRow
+                      ) : (
+                        <DragSourceAndDropTarget
+                          beginDrag={() => {
+                            this.props.onItemSelected(item);
+                            return {};
+                          }}
+                          canDrop={() =>
+                            canMoveSelectionToItem
+                              ? canMoveSelectionToItem(item)
+                              : true
+                          }
+                          drop={() => {
+                            onMoveSelectionToItem(item);
+                          }}
+                        >
+                          {({
+                            connectDragSource,
+                            connectDropTarget,
+                            isOver,
+                            canDrop,
+                          }) => {
+                            // Add an extra div because connectDropTarget/connectDragSource can
+                            // only be used on native elements
+                            const dropTarget = connectDropTarget(
+                              <div>
+                                {isOver && <DropIndicator canDrop={canDrop} />}
+                                {React.cloneElement(itemRow, {
```

Comment:
Prefer to use a function that returns the `<ItemRow` and take optional props as arguments. Avoid cloneElement in 99.9% of cases - it's making static analysis harder.

Reply:
After tinkering through the code, I have come to the conclusion, that to prevent dragging I have to return the <ItemRow ...> component directly and for dragging I have to wrap it in <DragSourceAndDropTarget ...> and then return it. 

Firstly, please correct me if I am wrong here. (i.e. if their is a possibility of preventing dragging and at the same time be wrapped inside <DragSourceAndDropTarget ...>).

And if my interpretation is correct, wrapping it will require additional prop of *connectIconDragSource* which can only be provided after wrapping it. 
So, I have come to following two solutions:
1. I can use the two separate instances of<ItemRow ...> for both cases (one inside the <DragSourceAndDropTarget ...> with *connectIconDragSource* prop and other without it). But, this will lead to duplication of code for <ItemRow ...> instance because all other props are same.
2. The solution proposed above of cloneElement, but as you correctly mentioned that it makes static analysis harder, so no to it.
3. Create separate renderItemRow function which takes props as arguments and return the <ItemRow ...>  instance.

So, which method to follow among these or any other approach you have come across.